samples/iothub-connect-to-eventhubs/IotHubConnection.cs

```diff
@@ -0,0 +1,355 @@
+            // is available.
+
+            var parsedTokens = new Dictionary<string, string>(ParseConnectionString(iotHubConnectionString), StringComparer.InvariantCultureIgnoreCase);
+            var hostName = parsedTokens["HostName"];
+            var sharedKeyName = parsedTokens["SharedAccessKeyName"];
+            var sharedKey = parsedTokens["SharedAccessKey"];
+            var iotHubName = hostName?.Substring(0, hostName.IndexOf('.'));
+
+            if ((string.IsNullOrEmpty(hostName)) || (string.IsNullOrEmpty(sharedKeyName)) || (string.IsNullOrEmpty(sharedKey)))
+            {
+                throw new ArgumentException("The IOT Hub connection string is not valid; it must contain the host, shared access key, and shared access key name.", nameof(iotHubConnectionString));
+            }
+
+            if (string.IsNullOrEmpty(iotHubName))
+            {
+                throw new ArgumentException("Unable to parse the IOT Hub name from the connection string.", nameof(iotHubConnectionString));
+            }
+
+            // Establish the IOT Hub connection via link to the necessary endpoint, which will trigger a redirect exception
+            // from which the Event Hubs connection string can be built.
+
+            var stopWatch = Stopwatch.StartNew();
+            var serviceEndpoint = new Uri($"{ AmqpConstants.SchemeAmqps }://{ hostName }/messages/events");
+            var connection = default(AmqpConnection);
+            var link = default(AmqpLink);
+            var eventHubsHost = default(string);
+
+            try
+            {
+                connection = await CreateAndOpenConnectionAsync(serviceEndpoint, iotHubName, sharedKeyName, sharedKey, timeout.Value).ConfigureAwait(false);
+                link = await CreateRedirectLinkAsync(connection, serviceEndpoint, timeout.Value.Subtract(stopWatch.Elapsed)).ConfigureAwait(false);
+
+                await link.OpenAsync(timeout.Value.Subtract(stopWatch.Elapsed)).ConfigureAwait(false);
+            }
+            catch (AmqpException ex)
+                when ((ex?.Error?.Condition.Value == AmqpErrorCode.LinkRedirect.Value) && (ex?.Error?.Info != null))
+            {
+                ex.Error.Info.TryGetValue("hostname", out eventHubsHost);
+            }
+            finally
+            {
+                stopWatch.Stop();
+
+                link?.Session?.SafeClose();
+                link?.SafeClose();
+                connection?.SafeClose();
+            }
+
+            // Attempt to assemble the Event Hubs connection string using the IOT Hub components.
+
+            if (string.IsNullOrEmpty(eventHubsHost))
+            {
+                throw new InvalidOperationException("The Event Hubs host was not returned by the IOT Hub service.");
+            }
+
+            return $"Endpoint=sb://{ eventHubsHost }/;EntityPath={ iotHubName };SharedAccessKeyName={ sharedKeyName };SharedAccessKey={ sharedKey }";
```

Comment:
Looks like this https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventhubs.eventhubsconnectionstringbuilder?view=azure-dotnet-legacy wasn't carried over to new SDK, but would be nice to have in this case.

Reply:
It's unfortunate that we need to produce a connection string here at all; generally speaking, we'd prefer to encourage using the namespace and Event Hub name with an identity credential.   In this case, we cannot reliably do so because IoT Hub owns the Event Hubs instances and, to my knowledge, doesn't support configuring RBAC grants for the child resources that it allocates/manages.

Not carrying the track one builder forward was an intentional decision to try to avoid going down the path of keep connection string use focused on exploring the library and not production usage.  In that same focus, we also wanted to consciously avoid making it tempting to extend the connection string with custom grammar for the client to parse and use, which was a path that track one had taken.